src/encoder.cpp

```diff
@@ -0,0 +1,169 @@
+#include "encoder.h"
+#include "reader_util.h"
+#include "scope_guard.h"
+#include <exception>
+
+#ifdef LCF_SUPPORT_ICU
+#   include <unicode/ucsdet.h>
+#   include <unicode/ucnv.h>
+#else
+#   ifdef _MSC_VER
+#		error MSVC builds require ICU
+#	endif
+#endif
+
+#ifdef _WIN32
+#   define WIN32_LEAN_AND_MEAN
+#   ifndef NOMINMAX
+#       define NOMINMAX
+#   endif
+#   include <windows.h>
+#else
+#   ifndef LCF_SUPPORT_ICU
+#       include <iconv.h>
+#   endif
+#   include <locale>
+#endif
+
+#if defined(__MORPHOS__) || defined(__amigaos4__)
+#define ICONV_CONST const
+#endif
+
+static std::string filterUtf8Compatible(std::string enc) {
+#ifdef LCF_SUPPORT_ICU
+	if (ucnv_compareNames(enc.c_str(), "UTF-8") == 0) {
+		return "";
+	}
+#endif
+	return enc;
+}
+
+Encoder::Encoder(std::string encoding)
+	: _encoding(filterUtf8Compatible(std::move(encoding)))
+{
+	Init();
+}
+
+Encoder::~Encoder() {
+	Reset();
+}
+
+void Encoder::Encode(std::string& str) {
+	if (_encoding.empty() || str.empty()) {
+		return;
+	}
+	Convert(str, _conv_runtime, _conv_storage);
+}
+
+void Encoder::Decode(std::string& str) {
+	if (_encoding.empty() || str.empty()) {
+		return;
+	}
+	Convert(str, _conv_storage, _conv_runtime);
+}
+
+void Encoder::Init() {
+	if (_encoding.empty()) {
+		return;
+	}
+	auto code_page = atoi(_encoding.c_str());
+	const auto& storage_encoding = code_page > 0
+		? ReaderUtil::CodepageToEncoding(code_page)
+		: _encoding;
+
+#ifdef LCF_SUPPORT_ICU
+	auto status = U_ZERO_ERROR;
+	constexpr auto runtime_encoding = "UTF-8";
+	auto conv_runtime = ucnv_open(runtime_encoding, &status);
+
+	if (conv_runtime == nullptr) {
+		fprintf(stderr, "liblcf:  ucnv_open() error for encoding \"%s\": %s\n", runtime_encoding, u_errorName(status));
+		throw std::runtime_error("ucnv_open() failed");
```

Comment:
some of our platforms are compiled with ``-fno-rtti and -fno-exceptions``.

The ucnv_open failed codepath is commonly taken on Player startup btw, as part of the heuristic to guess the encoding (ICU will suggest multiple languages and some of these languages are not part of our custom data file).